conjure-codegen/src/example_types/another/test_service.rs

```diff
@@ -0,0 +1,770 @@
+        if response.status() == conjure_http::private::http::StatusCode::NO_CONTENT {
+            Ok(Default::default())
+        } else {
+            conjure_http::private::json::client_from_reader(response.body_mut())
+                .map_err(conjure_http::private::Error::internal)
+        }
+    }
+    pub fn test_param(
+        &self,
+        auth_: &conjure_object::BearerToken,
+        dataset_rid: &conjure_object::ResourceIdentifier,
+    ) -> Result<Option<String>, conjure_http::private::Error> {
+        let mut request_ =
+            conjure_http::private::http::Request::new(conjure_http::client::Body::Empty);
+        *request_.method_mut() = conjure_http::private::http::Method::GET;
+        let path_ = format!(
+            "/catalog/datasets/{}/testParam",
+            conjure_http::private::percent_encode(
+                conjure_object::ToPlain::to_plain(&dataset_rid).as_bytes(),
+                conjure_http::private::PATH_SEGMENT_ENCODE_SET,
+            ),
+        );
+        *request_.uri_mut() = conjure_http::private::http::Uri::from_shared(path_.into())
+            .expect("URI should be valid");
+        request_.headers_mut().insert(
+            conjure_http::private::http::header::AUTHORIZATION,
+            conjure_http::private::http::header::HeaderValue::from_shared(
+                format!("Bearer {}", auth_.as_str()).into(),
+            )
+            .expect("bearer tokens are valid headers"),
+        );
+        request_.headers_mut().insert(
+            conjure_http::private::http::header::ACCEPT,
+            conjure_http::private::http::header::HeaderValue::from_static("application/json"),
+        );
+        let mut response = self.0.request(request_)?;
+        if response.status() == conjure_http::private::http::StatusCode::NO_CONTENT {
+            Ok(Default::default())
+        } else {
+            conjure_http::private::json::client_from_reader(response.body_mut())
+                .map_err(conjure_http::private::Error::internal)
+        }
+    }
+    pub fn test_query_params(
+        &self,
+        auth_: &conjure_object::BearerToken,
+        query: &str,
+        something: &conjure_object::ResourceIdentifier,
+        optional_middle: Option<&conjure_object::ResourceIdentifier>,
+        implicit: &conjure_object::ResourceIdentifier,
+        set_end: &std::collections::BTreeSet<String>,
+        optional_end: Option<&conjure_object::ResourceIdentifier>,
+    ) -> Result<i32, conjure_http::private::Error> {
+        let mut request_ =
+            conjure_http::private::http::Request::new(conjure_http::client::Body::Fixed(
+                conjure_http::private::json::to_vec(&query)
+                    .map_err(conjure_http::private::Error::internal)?,
+            ));
+        *request_.method_mut() = conjure_http::private::http::Method::POST;
+        let mut path_ = "/catalog/test-query-params".to_string();
+        path_.push_str("?different=");
```

Comment:
so this is good (and let's continue with this approach) but reading this method made me wonder whether we could instead just put them into the request and have the client library take care of putting them in the url?

fwiw we take this approach with python generator - just passing query params to requests via `params` argument: 
https://2.python-requests.org/en/master/user/quickstart/#passing-parameters-in-urls and https://github.com/palantir/conjure-python/blob/develop/conjure-python-core/src/main/java/com/palantir/conjure/python/poet/PythonEndpointDefinition.java#L133

Reply:
We have to generate the path anyway. I think ideally the http crate would gain a `UriBuilder` type that we could use here to simplify things.